x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/PruneJoinOnNullMatchingField.java

```diff
@@ -0,0 +1,80 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.core.expression.Expressions.isGuaranteedNull;
+import static org.elasticsearch.xpack.esql.plan.logical.join.JoinTypes.INNER;
+import static org.elasticsearch.xpack.esql.plan.logical.join.JoinTypes.LEFT;
+
+/**
+ * The rule matches a plan pattern having a Join on top of a Project and/or Eval. It then checks if the join's performed on a field which
+ * is aliased to null (in type or value); if that's the case, it prunes the join, replacing it with an Eval  - returning aliases to null
+ * for all the fields added in by the right side of the Join - plus a Project on top of it.
+ * The rule can apply on the coordinator already, but it's more likely to be effective on the data nodes, where null aliasing is inserted
+ * due to locally missing fields. This rule relies on that behavior -- see {@link ReplaceMissingFieldWithNull}.
+ */
+public class PruneJoinOnNullMatchingField extends OptimizerRules.OptimizerRule<Join> {
+
+    @Override
+    protected LogicalPlan rule(Join join) {
+        LogicalPlan plan = join;
+        var joinType = join.config().type();
+        if (joinType == INNER || joinType == LEFT) { // other types will have different replacement logic
+            AttributeMap.Builder<Expression> attributeMapBuilder = AttributeMap.builder();
+            loop: for (var child = join.left();; child = ((UnaryPlan) child).child()) { // cast is safe as both plans are UnaryPlans
+                switch (child) {
+                    case Project project -> project.projections().forEach(projection -> {
+                        if (projection instanceof Alias alias) {
+                            attributeMapBuilder.put(alias.toAttribute(), alias.child());
+                        }
+                    });
+                    case Eval eval -> eval.fields().forEach(alias -> attributeMapBuilder.put(alias.toAttribute(), alias.child()));
+                    default -> {
+                        break loop;
+                    }
+                }
+            }
+            for (var attr : AttributeSet.of(join.config().matchFields())) {
+                var resolved = attributeMapBuilder.build().resolve(attr);
+                if (resolved != null && isGuaranteedNull(resolved)) {
+                    plan = replaceJoin(join);
+                    break;
+                }
+            }
+        }
+        return plan;
+    }
+
+    private static LogicalPlan replaceJoin(Join join) {
+        var joinRightOutput = join.rightOutputFields();
+        if (joinRightOutput.isEmpty()) { // can be empty when the join key is null and the other right side entries pruned (by an agg)
+            return join.left();
+        }
+        List<Alias> aliases = new ArrayList<>(joinRightOutput.size());
+        // TODO: cache aliases by type, à la ReplaceMissingFieldWithNull#missingToNull (tho lookup indices won't have Ks of fields)
```

Comment:
Yep, let's re-use that code, maybe!

Not sure if lookup indices will have much fewer fields, I expect them to grow over time.

Reply:
> Not sure if lookup indices will have much fewer fields, I expect them to grow over time.

Assuming here, but I believe that the Ks of fields occurrence stems mostly from ECS and lookup indices would probably grow too, but controlled (as they're more "manual"?). However this can be an issue with "generic" indices joins, indeed. 

> Yep, let's re-use that code, maybe!

If we won't remodel our approach, ["simplifying this together with other optimizer rules, namely `ReplaceMissingFieldsWithNull` and `PropagateEvalFoldables`"](https://github.com/elastic/elasticsearch/pull/127583#pullrequestreview-2828611461), I'd add this in an immediate follow-up, as it'd touch `ReplaceMissingFieldsWithNull`... but can also do it here, if you feel strongly about this. :)

Reply:
No, I don't think this is a blocker. A follow-up is perfectly fine IMO. Thank you!

Reply:
> I'd add this in an immediate follow-up

I've pushed the change in this PR, after all.